Update CL-LDM-05-31-2023.md (#1)
captainsafia/csharplang: meetings/working-groups/collection-literals/CL-LDM-05-31-2023.md

[# C# LDM Collection Literals Agenda

## Core set of features for C# 12

### Target-typing to core collections and collections with classic collection initializers

The following are constructible collection types that can be used to construct collection literals. These are the valid target types for collection literals.

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

1. Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

2. Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

> 3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

Author: ```suggestion
3. Types with a suitable `Construct` method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.  It will also be the preferred way to create a `List<T>` as it is the most efficient form with lowest overhead.
```

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

> // Open question below about update vs throw semantics for duplicate keys.

Author: ```suggestion
// Open question below about update vs throw semantics for duplicate keys.  Will cover when we discuss dictionaries.
```

```

5. Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open questions on if there is a guarantee about concrete type instantiated.
```

6. Types that implement System.Collections.IEnumerable - collection initializers

```c#
HashSet<int> set = [1, 2, 3]; // equivalent to:

HashSet<int> set = new HashSet { 1, 2, 3 };

// We pass the capacity to the constructor if it takes one.
```

7. Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)

```c#
IEnumerable<int> values = [1, 2, 3]; // could be:

IEnumerable<int> values = new List<int> { 1, 2, 3 }; // or could be:
IEnumerable<int> valuies = new <>UnspeakableName { 1, 2, 3 };

// Open question below about this.
```

Open Questions:

> 1. Does LDM agree with the priority order and totality of the support constructible collection types?

Author: ```suggestion
1. Does LDM agree with the priority order and totality of the support constructible collection types?  Of note: this means that any type that is not in that list, and is not being updated (e.g. netfx 4.7.2 types) might not work with collection literals.  Should we have some way (e.g. an extension-method pathway) to be able to opt those legacy types in?
```

[# C# LDM Collection Literals Agenda

## Core set of features for C# 12

### Target-typing to core collections and collections with classic collection initializers

The following are constructible collection types that can be used to construct collection literals. These are the valid target types for collection literals.

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

> Single dimensional arrays (e.g. T[])

Author: ```suggestion
1. Single dimensional arrays (e.g. T[])
```

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

> Span types (`Span<T>` and `ReadOnlySpan<T>`)

Author: ```suggestion
2. Span types (`Span<T>` and `ReadOnlySpan<T>`)
```

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

> Types with a suitable Construct method.  This will be discussed below, but related heavily to immutable collections, and how to efficiently construct them.

Author: ```suggestion
3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.
```

> Types that implement System.Collections.IDictionary - dictionary collection initializers

Author: ```suggestion

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } };

// Open questions on this to be discussed below.  Specifically:
// a. syntax ambgituities
// b. guarantees about concrete types.
```

[# C# LDM Collection Literals Agenda

## Core set of features for C# 12

### Target-typing to core collections and collections with classic collection initializers

The following are constructible collection types that can be used to construct collection literals. These are the valid target types for collection literals.

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

> Actual translation of the literal to the corresponding type is defined below.

Author: ```suggestion
```

Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

> Types with a suitable Construct method

Author: ```suggestion
Types with a suitable Construct method.  This will be discussed below, but related heavily to immutable collections, and how to efficiently construct them.
```

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

1. Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

2. Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

> Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

Author: ```suggestion
Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open questions on if there is a guarantee about concrete type instantiated.
```

Types that implement System.Collections.IEnumerable - collection initializers

> Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)

Author: ````suggestion
// b. guarantees about concrete types.
```

Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)
Types that implement System.Collections.IEnumerable - collection initializers
Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)
````

[# C# LDM Collection Literals Agenda

## Core set of features for C# 12

### Target-typing to core collections and collections with classic collection initializers

The following are constructible collection types that can be used to construct collection literals. These are the valid target types for collection literals.

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

1. Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

2. Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#

> IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
> 
> IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } };

Author: ```suggestion
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

1. Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

2. Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.

// Open questions on this to be discussed below.  Specifically:
// a. syntax ambgituities

> // b. guarantees about concrete types.

Author: ```suggestion
```

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

> // Open questions on if there is a guarantee about concrete type instantiated.
> Types that implement System.Collections.IEnumerable - collection initializers
> Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)

Author: ````suggestion
// Open questions on if there is a guarantee about concrete type instantiated.
```

Types that implement System.Collections.IEnumerable - collection initializers
Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)
````

```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open questions on if there is a guarantee about concrete type instantiated.
```

> Types that implement System.Collections.IEnumerable - collection initializers
> Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)

Author: ````suggestion
6. Types that implement System.Collections.IEnumerable - collection initializers

```c#
HashSet<int> set = [1, 2, 3]; // equivalent to:

HashSet<int> set = new HashSet { 1, 2, 3 };
```

Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)
````

The list is in priority order. If a collection type belongs in multiple categories, the first is used.

1. Single dimensional arrays (e.g. T[])

```c#
int[] a = [1, 2, 3]; // equivalent to:
int[] a = new int[] { 1, 2, 3 };
```

2. Span types (`Span<T>` and `ReadOnlySpan<T>`)

```c#
Span<int> a = [1, 2, 3]; // equivalent to either:

Span<int> a = new int[] { 1, 2, 3 }; // or
Span<int> a = stackalloc int[] { 1, 2, 3 };
```

Which one is picked depends on compiler heuristics around stack space (and is part of the 'params span' WG outcomes), and the ref-safety rules around 'a'.

3. Types with a suitable Construct method.  This will be discussed below, but relates heavily to immutable collections, and how to efficiently construct them.

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

> Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

Author: ```suggestion
5. Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)
```

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open questions on if there is a guarantee about concrete type instantiated.
```

6. Types that implement System.Collections.IEnumerable - collection initializers

```c#
HashSet<int> set = [1, 2, 3]; // equivalent to:

> HashSet<int> set = new HashSet { 1, 2, 3 };

Author: ```suggestion
HashSet<int> set = new HashSet { 1, 2, 3 };

// We pass the capacity to the constructor if it takes one.
```

4. Types that implement System.Collections.IDictionary - dictionary collection initializers

```c#
Dictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:

Dictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open question below about update vs throw semantics for duplicate keys.
```

5. Interface types I<TKey, TValue> implemented by System.Collections.Generic.Dictionary<TKey, TValue> (e.g. IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>)

```c#
IDictionary<string, int> nameToAge = [ "Dustin": 42, "Cyrus": 43 ]; // possibly equivalent to:
IDictionary<string, int> nameToAge = new Dictionary<string, int> { { "Dustin": 42 }, { "Cyrus": 43 } }; 

// Open questions on if there is a guarantee about concrete type instantiated.
```

6. Types that implement System.Collections.IEnumerable - collection initializers

```c#
HashSet<int> set = [1, 2, 3]; // equivalent to:

HashSet<int> set = new HashSet { 1, 2, 3 };

// We pass the capacity to the constructor if it takes one.
```

> 
> 
> Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)
> 
> 
> https://github.com/dotnet/csharplang/blob/main/proposals/collection-literals.md#constructible-collection-types-target-types

Author: ````suggestion

7. Interface types I<T> implemented by System.Collections.Generic.List<T> (e.g. IEnumerable<T>, IList<T>, IReadOnlyList<T>)

```c#
IEnumerable<int> values = [1, 2, 3]; // could be:

IEnumerable<int> values = new List<int> { 1, 2, 3 }; // or could be:
IEnumerable<int> valuies = new <>UnspeakableName { 1, 2, 3 };

// Open question below about this.
```

````
